Enhancing the FTP protocol

RomanRII opened the pull request:
# What was done
* Modified `nxc/protocols/ftp.py`
  * Modified the `--ls` flag to allow for listing the current directory and sub-directories. Default now lists `.`. If an argument is provided, it will list the provided sub-directory
  * Added the `--get` flag to download a file on the server. If the file exists and is successfully downloaded, it will be written to the users cwd with the remote file's filename.
  * Added the `--put` flag to upload files onto the server. 
* Modified `nxc/protocols/ftp/proto_args.py` to reflect the added features
  * Modified the `--ls` flag to allow for a default directory listing (`.`) or use a provided directory
  * Added the `--get` and `--put` flags
* Modified [nxc/protocols/ftp.py#L83](https://github.com/Pennyw0rth/NetExec/blob/main/nxc/protocols/ftp.py#L83) to comply with [RFC 1635](https://www.rfc-editor.org/rfc/rfc1635.html)

# Test environment used
* During testing, different permissions were used to account for edge cases.
``` python3
from pyftpdlib.authorizers import DummyAuthorizer
from pyftpdlib.handlers import FTPHandler
from pyftpdlib.servers import FTPServer

def start_ftp_server():
    # Instantiate an authorizer object responsible for managing authentication and authorization
    authorizer = DummyAuthorizer()

    # Allow anonymous access with read-only permissions
    authorizer.add_anonymous("./FTP_Folder/", perm="erlw")

    # Instantiate an FTP handler object and link it to the authorizer
    handler = FTPHandler
    handler.authorizer = authorizer

    # Create the FTP server on the desired address and port, and link it to the handler
    server = FTPServer(("127.0.0.1", 21), handler)

    # Start the FTP server
    server.serve_forever()

if __name__ == "__main__":
    start_ftp_server()
```

RomanRII commented:
# RFC 1635 Compliance
![image](https://github.com/Pennyw0rth/NetExec/assets/74742067/0488c033-9474-403e-b23f-29dcb3f1c288)

# --ls 
## Standard usage
![image](https://github.com/Pennyw0rth/NetExec/assets/74742067/bf791e35-24ad-40ab-99c7-d76cba412f64)
## If permissions are not granted
![image](https://github.com/Pennyw0rth/NetExec/assets/74742067/a8707ec1-576c-4d5b-96f8-79684c54bac2)

# --get
## Standard usage
![image](https://github.com/Pennyw0rth/NetExec/assets/74742067/2780742d-0e45-4ae3-bce6-2c7c44b69a7a)
## If the file does not exist
![image](https://github.com/Pennyw0rth/NetExec/assets/74742067/9c0c2e47-37e7-424e-a39c-2e2ff08a167f)

# --put
## Standard usage
![image](https://github.com/Pennyw0rth/NetExec/assets/74742067/4fed417c-e8fb-4e6c-bebf-1c65ccde4e3a)
## If a local file does not exist
![image](https://github.com/Pennyw0rth/NetExec/assets/74742067/b3bc2d16-1694-4f57-b3e0-de5192e40217)
## Permissions error
![image](https://github.com/Pennyw0rth/NetExec/assets/74742067/d94c260d-a539-4958-8012-59ed4f1535aa)

RomanRII commented:
Couple things to mind about this merge, it currently doesn't allow the following:
## --put
* If a user attempts to upload a file like `--put example.txt .`, it will fail with `Failed to upload file. Response: (550 Is a directory.)`

The user will need to specify the filename or dir + filename to save to
`--put example.txt example.txt` or `--put example.txt Sub/Dir/example.txt`

If this current implementation is not prefered, let me know and I can make some adjustments to allow users to upload using either `.` or `Sub/` and upload to the directory using the local_name

Marshall-Hallenbeck commented:
@RomanRII can you add these new flags to the e2e tests commands file under the FTP section (tests/e2e_commands.txt)?

RomanRII commented:
@Marshall-Hallenbeck you got it. will get that done

RomanRII commented:
* Added new flags to ([tests/e2e_commands.txt](https://github.com/RomanRII/NetExec/blob/main/tests/e2e_commands.txt#L207))
* Added test file to test upload/download functions ([tests/data/test_file.txt](https://github.com/RomanRII/NetExec/blob/main/tests/data/test_file.txt))

NeffIsBack commented on nxc/protocols/ftp/proto_args.py:
```diff
@@ -3,5 +3,7 @@ def proto_args(parser, std_parser, module_parser):
     cgroup = ftp_parser.add_argument_group("FTP Access", "Options for enumerating your access")
-    cgroup.add_argument("--ls", action="store_true", help="List files in the directory")
+    cgroup.add_argument("--ls", metavar="DIRECTORY", nargs="?", const=".", help="List files in the directory")
+    cgroup.add_argument("--get", metavar="FILE", help="Download a file")
+    cgroup.add_argument("--put", metavar=("LOCAL_FILE", "REMOTE_FILE"), nargs=2, help="Upload a file")
```
Would be nice if you could add an example as in:
https://github.com/Pennyw0rth/NetExec/blob/66e6c95f0d4cabe4acabdfadaa69f8eb38bb056f/nxc/protocols/smb/proto_args.py#L75C45-L75C45

RomanRII replied:
Added: https://github.com/RomanRII/NetExec/commit/901b8dee80fc1109414c688f491f5ad4fdec9049

NeffIsBack commented on nxc/protocols/ftp.py:
```diff
@@ -80,16 +81,42 @@ def plaintext_login(self, username, password):
             else:
                 self.logger.success(f"{username}:{process_secret(password)}")
-
+ 
```
Please remove the space

RomanRII replied:
Removed space: https://github.com/RomanRII/NetExec/commit/f26b676c8ee94992006352737663352301ed201b

NeffIsBack reviewed (changes requested):
Really good PR! Clean code and very nice documentation of the changes! Love to see that

Marshall-Hallenbeck reviewed (approved):
Reviewed and tested this. It all works. Great code and feature, thank you!
